Module '"react"' has no exported member 'cache'

### Current Behavior

I'm trying to build a library using the latest react `cache` function.

I get the following error during build using rollup:
```ts
libs/client/react-query/src/lib/getQueryClient.ts:2:10 - error TS2305: Module '"react"' has no exported member 'cache'.

  1 | import { QueryClient } from '@tanstack/query-core';
> 2 | import { cache } from 'react';
    |          ^
  3 | 
  4 | export const getQueryClient = cache(() => new QueryClient());
  5 | 

Found 1 error.
```

### Expected Behavior

The issue is documented [here](https://github.com/vercel/next.js/issues/49707) and should be fixed by using the latest `@types/react` version.

The react types should not be fixed but parsed from the root `package.json`

https://github.com/nrwl/nx/blob/18c1be8b86024076aeccbf95e6282a1f480bbb18/packages/react/src/utils/versions.ts#L11

### GitHub Repo

https://github.com/sebpalluel/nx-examples

### Steps to Reproduce

1. `pnpm install`
2. `nx run react-query:build`


### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node   : 20.1.0
   OS     : darwin-arm64
   yarn   : 1.22.19
   
   nx                 : 16.4.0-beta.3
   @nx/js             : 16.4.0-beta.3
   @nx/jest           : 16.4.0-beta.3
   @nx/linter         : 16.4.0-beta.3
   @nx/workspace      : 16.4.0-beta.3
   @nx/angular        : 16.4.0-beta.3
   @nx/cypress        : 16.4.0-beta.3
   @nx/devkit         : 16.4.0-beta.3
   @nx/eslint-plugin  : 16.4.0-beta.3
   @nx/react          : 16.4.0-beta.3
   @nx/rollup         : 16.4.0-beta.3
   @nx/web            : 16.4.0-beta.3
   nx-cloud           : 16.0.5
   typescript         : 5.0.4
   ---------------------------------------
   Community plugins:
   @ngrx/component-store : 16.0.0
   @ngrx/effects         : 16.0.0
   @ngrx/entity          : 16.0.0
   @ngrx/router-store    : 16.0.0
   @ngrx/store           : 16.0.0
   @ngrx/store-devtools  : 16.0.0
   ---------------------------------------
   Local workspace plugins:
   	 @nx-example/react-query
```


### Failure Logs

_No response_

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

--- a/graph/client/src/app/feature-tasks/task-list.tsx
+++ b/graph/client/src/app/feature-tasks/task-list.tsx
@@ -44,7 +44,7 @@ function ProjectListItem({
 
       {project.error ? (
         <Tooltip
-          content={<TaskGraphErrorTooltip error={project.error} />}
+          content={(<TaskGraphErrorTooltip error={project.error} />) as any}
           openAction="click"
           strategy="fixed"
         >

--- a/package.json
+++ b/package.json
@@ -114,8 +114,8 @@
     "@types/marked": "^2.0.0",
     "@types/node": "18.16.9",
     "@types/prettier": "^2.6.2",
-    "@types/react": "18.0.25",
-    "@types/react-dom": "18.0.9",
+    "@types/react": "18.2.12",
+    "@types/react-dom": "18.2.5",
     "@types/semver": "^7.3.8",
     "@types/tar-stream": "^2.2.2",
     "@types/tmp": "^0.2.0",
@@ -349,4 +349,3 @@
     ]
   }
 }
-

--- a/packages/next/src/generators/library/library.ts
+++ b/packages/next/src/generators/library/library.ts
@@ -98,6 +98,15 @@ export async function libraryGenerator(host: Tree, rawOptions: Schema) {
         }
         return path;
       });
+      if (!json.compilerOptions) {
+        json.compilerOptions = {
+          types: [],
+        };
+      }
+      if (!json.compilerOptions.types) {
+        json.compilerOptions.types = [];
+      }
+      json.compilerOptions.types.push('next');
       return json;
     }
   );

--- a/packages/react/migrations.json
+++ b/packages/react/migrations.json
@@ -899,6 +899,23 @@
           "alwaysAddToPackageJson": false
         }
       }
+    },
+    "16.4.0-beta.7": {
+      "version": "16.4.0-beta.7",
+      "packages": {
+        "@types/react": {
+          "version": "18.2.12",
+          "alwaysAddToPackageJson": false
+        },
+        "@types/react-dom": {
+          "version": "18.2.5",
+          "alwaysAddToPackageJson": false
+        },
+        "@types/react-is": {
+          "version": "18.2.0",
+          "alwaysAddToPackageJson": false
+        }
+      }
     }
   }
 }

--- a/packages/react/src/utils/versions.ts
+++ b/packages/react/src/utils/versions.ts
@@ -8,9 +8,9 @@ export const reactDomVersion = '18.2.0';
 export const reactIsVersion = '18.2.0';
 export const swcLoaderVersion = '0.1.15';
 export const babelLoaderVersion = '^9.1.2';
-export const typesReactVersion = '18.0.28';
-export const typesReactDomVersion = '18.0.11';
-export const typesReactIsVersion = '17.0.3';
+export const typesReactVersion = '18.2.12';
+export const typesReactDomVersion = '18.2.5';
+export const typesReactIsVersion = '18.2.0';
 
 export const typesNodeVersion = '18.14.2';
 

